core/amber/src/main/scala/edu/uci/ics/amber/engine/common/rpc/AsyncRPCClient.scala

```diff
@@ -71,7 +72,14 @@ class AsyncRPCClient(controlOutputPort: ControlOutputPort, logger: WorkflowLogge
   def fulfillPromise(ret: ReturnInvocation): Unit = {
     if (unfulfilledPromises.contains(ret.originalCommandID)) {
       val p = unfulfilledPromises(ret.originalCommandID)
-      p.setValue(ret.controlReturn.asInstanceOf[p.returnType])
+
+      ret.controlReturn match {
+        case ControlException(msg) =>
```

Comment:
do we want to treat control exception as a regular message, or an exception of the promise? I understand the reason you want to treat it as an exception. 
But I'm not sure whether our Java side RPC client/server have this mechanism right now. If not, it might be better to make this also available in Java side., because the java control message handler might also throws an exception

Reply:
made java side to return Throwable and set Future's exception for now. A corresponding test case is added.